### PR TITLE
DOCSUP-13875: Document the replication_wait_for_inactive_replica_timeout setting

### PR DESCRIPTION
**Pull request by sevirov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:

- Documentation for #27931

### CODE REVIEW

File: docs/en/operations/settings/settings.md [hunk 1]
```diff
@@ -3457,3 +3457,27 @@ Possible values:
 -   1 — Projection optimization is obligatory.
 
 Default value: `0`.
+
+## replication_alter_partitions_sync {#replication-alter-partitions-sync}
+
+Allows to specify waiting for actions to manipulate the partitions.
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3463]:
```suggestion
Allows to specify the delay for actions that manipulate the partitions.
```

File: docs/en/operations/settings/settings.md [hunk 2]
```diff
@@ -3457,3 +3457,27 @@ Possible values:
...
+
+Possible values:
+
+-   0 — Do not wait.
+-   1 — To wait for own execution.
+-   2 — To wait for everyone.
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3469]:
```suggestion
-   0 — Do not wait. Means that the next action can start any time, not waiting for the previous action to be completed.
-   1 — Wait for own execution. Means that the next action on the partition starts only after the previous action is completed on this partition.
-   2 — Wait for everyone. Means that the next action starts only after the previous action is completed on all partitions.
```

File: docs/en/operations/settings/settings.md [hunk 3]
```diff
@@ -3457,3 +3457,27 @@ Possible values:
...
+
+Default value: `1`.
+
+## replication_wait_for_inactive_replica_timeout {#replication-wait-for-inactive-replica-timeout}
+
+Allows to specify how long (in seconds) to wait for inactive replicas to execute [ALTER](../../sql-reference/statements/alter/index.md), [OPTIMIZE](../../sql-reference/statements/optimize.md) or [TRUNCATE](../../sql-reference/statements/truncate.md) queries.
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3475]:
```suggestion
Specifies how long (in seconds) to wait for inactive replicas to execute [ALTER](../../sql-reference/statements/alter/index.md), [OPTIMIZE](../../sql-reference/statements/optimize.md) or [TRUNCATE](../../sql-reference/statements/truncate.md) queries.
```

File: docs/en/operations/settings/settings.md [hunk 4]
```diff
@@ -3457,3 +3457,27 @@ Possible values:
...
+
+Possible values:
+
+-   0 — Do not wait.
+-   Negative integer — To wait for unlimited time.
+-   Positive integer — The number of seconds of waiting.
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3481]:
```suggestion
-   Negative integer — Wait for unlimited time.
-   Positive integer — The number of seconds to wait.
```

File: docs/en/sql-reference/statements/alter/index.md [hunk 5]
```diff
@@ -43,7 +43,11 @@ Entries for finished mutations are not deleted right away (the number of preserv
 
 For non-replicated tables, all `ALTER` queries are performed synchronously. For replicated tables, the query just adds instructions for the appropriate actions to `ZooKeeper`, and the actions themselves are performed as soon as possible. However, the query can wait for these actions to be completed on all the replicas.
 
-For `ALTER ... ATTACH|DETACH|DROP` queries, you can use the `replication_alter_partitions_sync` setting to set up waiting. Possible values: `0` – do not wait; `1` – only wait for own execution (default); `2` – wait for all.
+For `ALTER ... ATTACH|DETACH|DROP` queries, you can use the [replication_alter_partitions_sync](../../../operations/settings/settings.md#replication-alter-partitions-sync) setting to set up waiting.
 
 For `ALTER TABLE ... UPDATE|DELETE` queries the synchronicity is defined by the [mutations_sync](../../../operations/settings/settings.md#mutations_sync) setting.
 
+You can to specify how long (in seconds) to wait for inactive replicas to execute all `ALTER` queries by the [replication_wait_for_inactive_replica_timeout seconds](../../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```
Comment (olgarev) [inline docs/en/sql-reference/statements/alter/index.md:50]:
```suggestion
You can to specify how long (in seconds) to wait for inactive replicas to execute all `ALTER` queries with the [replication_wait_for_inactive_replica_timeout seconds](../../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```

File: docs/en/sql-reference/statements/alter/index.md [hunk 6]
```diff
@@ -43,7 +43,11 @@ Entries for finished mutations are not deleted right away (the number of preserv
...
+
+!!! warning "Warning"
+    For all `ALTER` queries, if the `replication_alter_partitions_sync` setting is `2` and some replicas are not active for more than the time specified by the `replication_wait_for_inactive_replica_timeout seconds` setting, then an exception `UNFINISHED` will be thrown.
```
Comment (olgarev) [inline docs/en/sql-reference/statements/alter/index.md:53]:
```suggestion
    For all `ALTER` queries, if `replication_alter_partitions_sync = 2` and some replicas are not active for more than the time, specified in the `replication_wait_for_inactive_replica_timeout seconds` setting, then an exception `UNFINISHED` is thrown.
```

File: docs/en/sql-reference/statements/optimize.md [hunk 7]
```diff
@@ -25,6 +25,10 @@ When `OPTIMIZE` is used with the [ReplicatedMergeTree](../../engines/table-engin
 -   If you specify `FINAL`, optimization is performed even when all the data is already in one part. Also merge is forced even if concurrent merges are performed.
 -   If you specify `DEDUPLICATE`, then completely identical rows (unless by-clause is specified) will be deduplicated (all columns are compared), it makes sense only for the MergeTree engine.
 
+You can to specify how long (in seconds) to wait for inactive replicas to execute `OPTIMIZE` queries by the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
+
+!!! warning "Warning"
+    If the [replication_alter_partitions_sync](../../operations/settings/settings.md#replication-alter-partitions-sync) setting is `2` and some replicas are not active for more than the time specified by the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting, then an exception `UNFINISHED` will be thrown.
```
Comment (olgarev) [inline docs/en/sql-reference/statements/optimize.md:31]:
```suggestion
    If the [replication_alter_partitions_sync](../../operations/settings/settings.md#replication-alter-partitions-sync) is set to `2` and some replicas are not active for more than the time, specified by the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting, then an exception `UNFINISHED` is thrown.
```

File: docs/en/sql-reference/statements/truncate.md [hunk 8]
```diff
@@ -12,3 +12,8 @@ TRUNCATE TABLE [IF EXISTS] [db.]name [ON CLUSTER cluster]
 Removes all data from a table. When the clause `IF EXISTS` is omitted, the query returns an error if the table does not exist.
 
 The `TRUNCATE` query is not supported for [View](../../engines/table-engines/special/view.md), [File](../../engines/table-engines/special/file.md), [URL](../../engines/table-engines/special/url.md), [Buffer](../../engines/table-engines/special/buffer.md) and [Null](../../engines/table-engines/special/null.md) table engines.
+
+You can to specify how long (in seconds) to wait for inactive replicas to execute `TRUNCATE` queries by the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```
Comment (olgarev) [inline docs/en/sql-reference/statements/truncate.md:16]:
```suggestion
You can to specify how long (in seconds) to wait for inactive replicas to execute `TRUNCATE` queries with the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```

File: docs/en/sql-reference/statements/truncate.md [hunk 9]
```diff
@@ -12,3 +12,8 @@ TRUNCATE TABLE [IF EXISTS] [db.]name [ON CLUSTER cluster]
...
+
+!!! warning "Warning"
+    If the [replication_alter_partitions_sync](../../operations/settings/settings.md#replication-alter-partitions-sync) setting is `2` and some replicas are not active for more than the time specified by the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting, then an exception `UNFINISHED` will be thrown.
```
Comment (olgarev) [inline docs/en/sql-reference/statements/truncate.md:19]:
```suggestion
    If the [replication_alter_partitions_sync](../../operations/settings/settings.md#replication-alter-partitions-sync) is set to `2` and some replicas are not active for more than the time, specified by the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting, then an exception `UNFINISHED` is thrown.
```

File: docs/en/operations/settings/settings.md [hunk 10]
```diff
@@ -3458,6 +3458,30 @@ Possible values:
...
 
+## replication_alter_partitions_sync {#replication-alter-partitions-sync}
+
+Determines how actions that manipulate partitions are executed.
+
+Possible values:
+
+-   0 — Do not wait. Means that the next action can start any time, not waiting for the previous action to be completed.
+-   1 — Wait for own execution. Means that the next action on the partition starts only after the previous action is completed on this partition.
+-   2 — Wait for everyone. Means that the next action starts only after the previous action is completed on all partitions.
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3469]:
```suggestion
-   0 — Do not wait. The next action can start any time, not waiting for the previous action to be completed.
-   1 — Wait for own execution. The next action on the partition starts only after the previous action is completed on this partition.
-   2 — Wait for everyone. The next action starts only after the previous action is completed on all partitions.
```

File: docs/en/sql-reference/statements/alter/index.md [hunk 11]
```diff
@@ -43,7 +43,11 @@ Entries for finished mutations are not deleted right away (the number of preserv
 
 For non-replicated tables, all `ALTER` queries are performed synchronously. For replicated tables, the query just adds instructions for the appropriate actions to `ZooKeeper`, and the actions themselves are performed as soon as possible. However, the query can wait for these actions to be completed on all the replicas.
 
-For `ALTER ... ATTACH|DETACH|DROP` queries, you can use the `replication_alter_partitions_sync` setting to set up waiting. Possible values: `0` – do not wait; `1` – only wait for own execution (default); `2` – wait for all.
+For `ALTER ... ATTACH|DETACH|DROP` queries, you can use the [replication_alter_partitions_sync](../../../operations/settings/settings.md#replication-alter-partitions-sync) setting to set up waiting.
 
 For `ALTER TABLE ... UPDATE|DELETE` queries the synchronicity is defined by the [mutations_sync](../../../operations/settings/settings.md#mutations_sync) setting.
 
+You can to specify how long (in seconds) to wait for inactive replicas to execute all `ALTER` queries with the [replication_wait_for_inactive_replica_timeout seconds](../../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```
Comment (olgarev) [inline docs/en/sql-reference/statements/alter/index.md:50]:
```suggestion
You can specify how long (in seconds) to wait for inactive replicas to execute all `ALTER` queries with the [replication_wait_for_inactive_replica_timeout seconds](../../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```

File: docs/en/sql-reference/statements/optimize.md [hunk 12]
```diff
@@ -25,6 +25,10 @@ When `OPTIMIZE` is used with the [ReplicatedMergeTree](../../engines/table-engin
 -   If you specify `FINAL`, optimization is performed even when all the data is already in one part. Also merge is forced even if concurrent merges are performed.
 -   If you specify `DEDUPLICATE`, then completely identical rows (unless by-clause is specified) will be deduplicated (all columns are compared), it makes sense only for the MergeTree engine.
 
+You can to specify how long (in seconds) to wait for inactive replicas to execute `OPTIMIZE` queries by the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```
Comment (olgarev) [inline docs/en/sql-reference/statements/optimize.md:28]:
```suggestion
You can specify how long (in seconds) to wait for inactive replicas to execute `OPTIMIZE` queries by the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```

File: docs/en/sql-reference/statements/truncate.md [hunk 13]
```diff
@@ -12,3 +12,8 @@ TRUNCATE TABLE [IF EXISTS] [db.]name [ON CLUSTER cluster]
 Removes all data from a table. When the clause `IF EXISTS` is omitted, the query returns an error if the table does not exist.
 
 The `TRUNCATE` query is not supported for [View](../../engines/table-engines/special/view.md), [File](../../engines/table-engines/special/file.md), [URL](../../engines/table-engines/special/url.md), [Buffer](../../engines/table-engines/special/buffer.md) and [Null](../../engines/table-engines/special/null.md) table engines.
+
+You can to specify how long (in seconds) to wait for inactive replicas to execute `TRUNCATE` queries with the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```
Comment (olgarev) [inline docs/en/sql-reference/statements/truncate.md:16]:
```suggestion
You can specify how long (in seconds) to wait for inactive replicas to execute `TRUNCATE` queries with the [replication_wait_for_inactive_replica_timeout seconds](../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
```

File: docs/ru/operations/settings/settings.md [hunk 14]
```diff
@@ -3275,6 +3275,30 @@ SETTINGS index_granularity = 8192 │
...
 
+## replication_alter_partitions_sync {#replication-alter-partitions-sync}
+
+Определяет, как выполняются действия, управляющие партициями.
+
+Возможные значения:
+
+-   0 — не ждать. Означает, что следующее действие может начаться в любое время, не дожидаясь завершения предыдущего действия.
+-   1 — ждать выполнения действия на одной партиции. Означает, что следующее действие на партиции начинается только после завершения предыдущего действия на ней.
+-   2 — ждать выполнения действия на всех партициях. Означает, что следующее действие начинается только после завершения предыдущего действия на всех партициях.
```
Comment (olgarev) [inline docs/ru/operations/settings/settings.md:3286]:
```suggestion
-   0 — не ждать. Означает, что следующее действие может начаться в любое время, не дожидаясь завершения предыдущего действия.
-   1 — ждать выполнения действия на одной партиции. Означает, что следующее действие на партиции начинается только после завершения предыдущего действия на этой партиции.
-   2 — ждать выполнения действия на всех партициях. Означает, что следующее действие начинается только после завершения предыдущего действия на всех партициях.
```

File: docs/ru/operations/settings/settings.md [hunk 15]
```diff
@@ -3275,6 +3275,30 @@ SETTINGS index_granularity = 8192 │
...
+
+Указывает время ожидания (в секундах) выполнения запросов [ALTER](../../sql-reference/statements/alter/index.md), [OPTIMIZE](../../sql-reference/statements/optimize.md) или [TRUNCATE](../../sql-reference/statements/truncate.md) для неактивных реплик.
+
+Возможные значения:
+
+-   0 — не ждать.
+-   Отрицательное целое число — ждать неограниченное время.
+-   Положительное целое число — установить соответствующее количество секунд ожидания.
+
+Значение по умолчанию: `120`.
```
Comment (olgarev) [inline docs/ru/operations/settings/settings.md:3300]:
```suggestion
Значение по умолчанию: `120` секунд.
```

File: docs/en/operations/settings/settings.md [hunk 16]
```diff
@@ -3458,6 +3458,30 @@ Possible values:
...
+
+Specifies how long (in seconds) to wait for inactive replicas to execute [ALTER](../../sql-reference/statements/alter/index.md), [OPTIMIZE](../../sql-reference/statements/optimize.md) or [TRUNCATE](../../sql-reference/statements/truncate.md) queries.
+
+Possible values:
+
+-   0 — Do not wait.
+-   Negative integer — Wait for unlimited time.
+-   Positive integer — The number of seconds to wait.
+
+Default value: `120`.
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3483]:
```suggestion
Default value: `120` seconds.
```

**Comment by sevirov** [comment]
@tavplubix Please, review this PR.

File: docs/en/operations/settings/settings.md [hunk 17]
```diff
@@ -3458,6 +3458,30 @@ Possible values:
...
 
+## replication_alter_partitions_sync {#replication-alter-partitions-sync}
+
+Determines how actions that manipulate partitions are executed.
+
+Possible values:
+
+-   0 — Do not wait. The next action can start any time, not waiting for the previous action to be completed.
+-   1 — Wait for own execution. The next action on the partition starts only after the previous action is completed on this partition.
+-   2 — Wait for everyone. The next action starts only after the previous action is completed on all partitions.
```
Comment (tavplubix) [inline docs/en/operations/settings/settings.md:3469]: What is "the next action" and "the previous action"?

File: docs/en/sql-reference/statements/alter/index.md [hunk 18]
```diff
@@ -43,7 +43,11 @@ Entries for finished mutations are not deleted right away (the number of preserv
...
 
-For `ALTER ... ATTACH|DETACH|DROP` queries, you can use the `replication_alter_partitions_sync` setting to set up waiting. Possible values: `0` – do not wait; `1` – only wait for own execution (default); `2` – wait for all.
+For `ALTER ... ATTACH|DETACH|DROP` queries, you can use the [replication_alter_partitions_sync](../../../operations/settings/settings.md#replication-alter-partitions-sync) setting to set up waiting.
 
 For `ALTER TABLE ... UPDATE|DELETE` queries the synchronicity is defined by the [mutations_sync](../../../operations/settings/settings.md#mutations_sync) setting.
 
+You can specify how long (in seconds) to wait for inactive replicas to execute all `ALTER` queries with the [replication_wait_for_inactive_replica_timeout seconds](../../../operations/settings/settings.md#replication-wait-for-inactive-replica-timeout) setting.
+
+!!! warning "Warning"
+    For all `ALTER` queries, if `replication_alter_partitions_sync = 2` and some replicas are not active for more than the time, specified in the `replication_wait_for_inactive_replica_timeout seconds` setting, then an exception `UNFINISHED` is thrown.
```
Comment (tavplubix) [inline docs/en/sql-reference/statements/alter/index.md:53]: It's rather a note than a warning

**Review by tavplubix — CHANGES_REQUESTED** [review]

File: docs/en/sql-reference/statements/optimize.md [hunk 19]
```diff
@@ -18,13 +18,17 @@ OPTIMIZE TABLE [db.]name [ON CLUSTER cluster] [PARTITION partition | PARTITION I
 
 The `OPTMIZE` query is supported for [MergeTree](../../engines/table-engines/mergetree-family/mergetree.md) family, the [MaterializedView](../../engines/table-engines/special/materializedview.md) and the [Buffer](../../engines/table-engines/special/buffer.md) engines. Other table engines aren’t supported.
 
-When `OPTIMIZE` is used with the [ReplicatedMergeTree](../../engines/table-engines/mergetree-family/replication.md) family of table engines, ClickHouse creates a task for merging and waits for execution on all nodes (if the `replication_alter_partitions_sync` setting is enabled).
+When `OPTIMIZE` is used with the [ReplicatedMergeTree](../../engines/table-engines/mergetree-family/replication.md) family of table engines, ClickHouse creates a task for merging and waits for execution on all nodes (if the [replication_alter_partitions_sync](../../operations/settings/settings.md#replication-alter-partitions-sync) setting is enabled).
```
Comment (tavplubix) [inline docs/en/sql-reference/statements/optimize.md:21]: It waits for all replicas if `replication_alter_partitions_sync` is set to 2 (and waits for current replica if it is set to 1)

**Review by tavplubix — APPROVED** [review]

**Pull request merged**